Outerloop test builds broken

Noticed this in some of the jit optional pipelines but it's in outerloop too.

https://dev.azure.com/dnceng-public/public/_build/results?buildId=657460&view=results

```
/__w/1/s/src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs(71,85): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj] [/__w/1/s/src/tests/build.proj]
##[error]src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs(71,85): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj]
/__w/1/s/src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs(85,31): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj] [/__w/1/s/src/tests/build.proj]
##[error]src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs(85,31): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj]
/__w/1/s/src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs(170,50): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj] [/__w/1/s/src/tests/build.proj]
##[error]src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs(170,50): error CS0436: The type 'TypeName' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs' conflicts with the imported type 'TypeName' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj]
/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/AssemblyNameInfo.cs(182,23): error CS0436: The type 'AssemblyNameInfo' in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/AssemblyNameInfo.cs' conflicts with the imported type 'AssemblyNameInfo' in 'System.Reflection.Metadata, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Using the type defined in '/__w/1/s/src/libraries/Common/src/System/Reflection/Metadata/AssemblyNameInfo.cs'. [/__w/1/s/src/tests/ilverify/ILVerificationTests.csproj] [/__w/1/s/src/tests/build.proj]
...etc..
```
Looks like it all may be ILVerify related.

--- a/eng/Version.Details.xml
+++ b/eng/Version.Details.xml
@@ -282,43 +282,43 @@
       <Uri>https://github.com/dotnet/llvm-project</Uri>
       <Sha>26f8c30340764cfa7fa9090dc01a36c222bf09c1</Sha>
     </Dependency>
-    <Dependency Name="Microsoft.NETCore.App.Runtime.win-x64" Version="9.0.0-preview.4.24219.3">
+    <Dependency Name="Microsoft.NETCore.App.Runtime.win-x64" Version="9.0.0-preview.4.24227.6">
       <Uri>https://github.com/dotnet/runtime</Uri>
-      <Sha>56610095196ac12a397b1acd00835db4d86849b9</Sha>
+      <Sha>b8fe1d0ccb60926d6e7ef5d42e2398c178c9ee5c</Sha>
     </Dependency>
-    <Dependency Name="runtime.native.System.IO.Ports" Version="9.0.0-preview.4.24219.3">
+    <Dependency Name="runtime.native.System.IO.Ports" Version="9.0.0-preview.4.24227.6">
       <Uri>https://github.com/dotnet/runtime</Uri>
-      <Sha>56610095196ac12a397b1acd00835db4d86849b9</Sha>
+      <Sha>b8fe1d0ccb60926d6e7ef5d42e2398c178c9ee5c</Sha>
     </Dependency>
-    <Dependency Name="Microsoft.NETCore.ILAsm" Version="9.0.0-preview.4.24219.3">
+    <Dependency Name="Microsoft.NETCore.ILAsm" Version="9.0.0-preview.4.24227.6">
       <Uri>https://github.com/dotnet/runtime</Uri>
-      <Sha>56610095196ac12a397b1acd00835db4d86849b9</Sha>
+      <Sha>b8fe1d0ccb60926d6e7ef5d42e2398c178c9ee5c</Sha>
     </Dependency>
-    <Dependency Name="Microsoft.NET.Sdk.IL" Version="9.0.0-preview.4.24219.3">
+    <Dependency Name="Microsoft.NET.Sdk.IL" Version="9.0.0-preview.4.24227.6">
       <Uri>https://github.com/dotnet/runtime</Uri>
-      <Sha>56610095196ac12a397b1acd00835db4d86849b9</Sha>
+      <Sha>b8fe1d0ccb60926d6e7ef5d42e2398c178c9ee5c</Sha>
     </Dependency>
-    <Dependency Name="System.Text.Json" Version="9.0.0-preview.4.24219.3">
+    <Dependency Name="System.Text.Json" Version="9.0.0-preview.4.24227.6">
       <Uri>https://github.com/dotnet/runtime</Uri>
-      <Sha>56610095196ac12a397b1acd00835db4d86849b9</Sha>
+      <Sha>b8fe1d0ccb60926d6e7ef5d42e2398c178c9ee5c</Sha>
     </Dependency>
     <!-- Intermediate is necessary for source build. -->
-    <Dependency Name="Microsoft.SourceBuild.Intermediate.runtime.linux-x64" Version="9.0.0-preview.4.24219.3">
+    <Dependency Name="Microsoft.SourceBuild.Intermediate.runtime.linux-x64" Version="9.0.0-preview.4.24227.6">
       <Uri>https://github.com/dotnet/runtime</Uri>
-      <Sha>56610095196ac12a397b1acd00835db4d86849b9</Sha>
+      <Sha>b8fe1d0ccb60926d6e7ef5d42e2398c178c9ee5c</Sha>
       <SourceBuild RepoName="runtime" ManagedOnly="false" />
     </Dependency>
-    <Dependency Name="Microsoft.DotNet.ILCompiler" Version="9.0.0-preview.4.24219.3">
+    <Dependency Name="Microsoft.DotNet.ILCompiler" Version="9.0.0-preview.4.24227.6">
       <Uri>https://github.com/dotnet/runtime</Uri>
-      <Sha>56610095196ac12a397b1acd00835db4d86849b9</Sha>
+      <Sha>b8fe1d0ccb60926d6e7ef5d42e2398c178c9ee5c</Sha>
     </Dependency>
-    <Dependency Name="System.Reflection.Metadata" Version="9.0.0-preview.4.24219.3">
+    <Dependency Name="System.Reflection.Metadata" Version="9.0.0-preview.4.24227.6">
       <Uri>https://github.com/dotnet/runtime</Uri>
-      <Sha>56610095196ac12a397b1acd00835db4d86849b9</Sha>
+      <Sha>b8fe1d0ccb60926d6e7ef5d42e2398c178c9ee5c</Sha>
     </Dependency>
-    <Dependency Name="System.Reflection.MetadataLoadContext" Version="9.0.0-preview.4.24219.3">
+    <Dependency Name="System.Reflection.MetadataLoadContext" Version="9.0.0-preview.4.24227.6">
       <Uri>https://github.com/dotnet/runtime</Uri>
-      <Sha>56610095196ac12a397b1acd00835db4d86849b9</Sha>
+      <Sha>b8fe1d0ccb60926d6e7ef5d42e2398c178c9ee5c</Sha>
     </Dependency>
     <Dependency Name="Microsoft.DotNet.XHarness.TestRunners.Common" Version="9.0.0-prerelease.24224.1">
       <Uri>https://github.com/dotnet/xharness</Uri>

--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -104,10 +104,10 @@
     <!-- NuGet dependencies -->
     <NuGetBuildTasksPackVersion>6.0.0-preview.1.102</NuGetBuildTasksPackVersion>
     <!-- Installer dependencies -->
-    <MicrosoftNETCoreAppRuntimewinx64Version>9.0.0-preview.4.24219.3</MicrosoftNETCoreAppRuntimewinx64Version>
+    <MicrosoftNETCoreAppRuntimewinx64Version>9.0.0-preview.4.24227.6</MicrosoftNETCoreAppRuntimewinx64Version>
     <MicrosoftExtensionsDependencyModelVersion>6.0.0</MicrosoftExtensionsDependencyModelVersion>
     <!-- ILAsm dependencies -->
-    <MicrosoftNETCoreILAsmVersion>9.0.0-preview.4.24219.3</MicrosoftNETCoreILAsmVersion>
+    <MicrosoftNETCoreILAsmVersion>9.0.0-preview.4.24227.6</MicrosoftNETCoreILAsmVersion>
     <!-- Libraries dependencies -->
     <MicrosoftBclAsyncInterfacesVersion>6.0.0</MicrosoftBclAsyncInterfacesVersion>
     <MicrosoftBclHashCodeVersion>1.1.1</MicrosoftBclHashCodeVersion>
@@ -120,18 +120,18 @@
     <SystemIOFileSystemAccessControlVersion>5.0.0</SystemIOFileSystemAccessControlVersion>
     <SystemMemoryVersion>4.5.5</SystemMemoryVersion>
     <SystemReflectionMetadataVersion>9.0.0-preview.4.24227.6</SystemReflectionMetadataVersion>
-    <SystemReflectionMetadataLoadContextVersion>9.0.0-preview.4.24219.3</SystemReflectionMetadataLoadContextVersion>
+    <SystemReflectionMetadataLoadContextVersion>9.0.0-preview.4.24227.6</SystemReflectionMetadataLoadContextVersion>
     <SystemSecurityAccessControlVersion>6.0.0</SystemSecurityAccessControlVersion>
     <SystemSecurityCryptographyCngVersion>5.0.0</SystemSecurityCryptographyCngVersion>
     <SystemSecurityCryptographyOpenSslVersion>5.0.0</SystemSecurityCryptographyOpenSslVersion>
     <SystemSecurityPrincipalWindowsVersion>5.0.0</SystemSecurityPrincipalWindowsVersion>
     <SystemSecurityPermissionsVersion>7.0.0</SystemSecurityPermissionsVersion>
-    <SystemTextJsonVersion>9.0.0-preview.4.24219.3</SystemTextJsonVersion>
+    <SystemTextJsonVersion>9.0.0-preview.4.24227.6</SystemTextJsonVersion>
     <SystemRuntimeCompilerServicesUnsafeVersion>6.0.0</SystemRuntimeCompilerServicesUnsafeVersion>
     <SystemThreadingAccessControlVersion>7.0.0</SystemThreadingAccessControlVersion>
     <SystemThreadingTasksExtensionsVersion>4.5.4</SystemThreadingTasksExtensionsVersion>
     <SystemValueTupleVersion>4.5.0</SystemValueTupleVersion>
-    <runtimenativeSystemIOPortsVersion>9.0.0-preview.4.24219.3</runtimenativeSystemIOPortsVersion>
+    <runtimenativeSystemIOPortsVersion>9.0.0-preview.4.24227.6</runtimenativeSystemIOPortsVersion>
     <!-- Keep toolset versions in sync with dotnet/msbuild and dotnet/sdk -->
     <SystemCollectionsImmutableToolsetVersion>8.0.0</SystemCollectionsImmutableToolsetVersion>
     <SystemTextJsonToolsetVersion>8.0.0</SystemTextJsonToolsetVersion>

--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -119,7 +119,7 @@
     <SystemDrawingCommonVersion>8.0.0</SystemDrawingCommonVersion>
     <SystemIOFileSystemAccessControlVersion>5.0.0</SystemIOFileSystemAccessControlVersion>
     <SystemMemoryVersion>4.5.5</SystemMemoryVersion>
-    <SystemReflectionMetadataVersion>9.0.0-preview.4.24219.3</SystemReflectionMetadataVersion>
+    <SystemReflectionMetadataVersion>9.0.0-preview.4.24227.6</SystemReflectionMetadataVersion>
     <SystemReflectionMetadataLoadContextVersion>9.0.0-preview.4.24219.3</SystemReflectionMetadataLoadContextVersion>
     <SystemSecurityAccessControlVersion>6.0.0</SystemSecurityAccessControlVersion>
     <SystemSecurityCryptographyCngVersion>5.0.0</SystemSecurityCryptographyCngVersion>
@@ -210,7 +210,7 @@
     <!-- Mono Cecil -->
     <MicrosoftDotNetCecilVersion>0.11.4-alpha.24222.1</MicrosoftDotNetCecilVersion>
     <!-- ILCompiler -->
-    <MicrosoftDotNetILCompilerVersion>9.0.0-preview.4.24219.3</MicrosoftDotNetILCompilerVersion>
+    <MicrosoftDotNetILCompilerVersion>9.0.0-preview.4.24227.6</MicrosoftDotNetILCompilerVersion>
     <!-- ICU -->
     <MicrosoftNETCoreRuntimeICUTransportVersion>9.0.0-preview.5.24223.2</MicrosoftNETCoreRuntimeICUTransportVersion>
     <!-- MsQuic -->

--- a/src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs
+++ b/src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs
@@ -29,6 +29,8 @@ namespace System.Reflection
 {
     internal partial struct TypeNameParser
     {
+        private static readonly TypeNameParseOptions s_typeNameParseOptions = new() { MaxNodes = int.MaxValue };
+
         private ModuleDesc _module;
         private bool _throwIfNotFound;
         private Func<ModuleDesc, string, MetadataType> _canonResolver;
@@ -36,7 +38,7 @@ namespace System.Reflection
         public static TypeDesc ResolveType(ModuleDesc module, string name, bool throwIfNotFound,
             Func<ModuleDesc, string, MetadataType> canonResolver)
         {
-            if (!TypeName.TryParse(name.AsSpan(), out TypeName parsed))
+            if (!TypeName.TryParse(name.AsSpan(), out TypeName parsed, s_typeNameParseOptions))
             {
                 ThrowHelper.ThrowTypeLoadException(name, module);
             }

--- a/src/coreclr/tools/ILVerification/ILVerification.projitems
+++ b/src/coreclr/tools/ILVerification/ILVerification.projitems
@@ -66,30 +66,6 @@
     <Compile Include="$(ToolsCommonPath)TypeSystem\Common\Utilities\CustomAttributeTypeNameParser.cs">
       <Link>Utilities\CustomAttributeTypeNameParser.cs</Link>
     </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\HexConverter.cs">
-      <Link>Utilities\HexConverter.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\AssemblyNameFormatter.cs">
-      <Link>Utilities\AssemblyNameFormatter.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\AssemblyNameParser.cs">
-      <Link>Utilities\AssemblyNameParser.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\Metadata\AssemblyNameInfo.cs">
-      <Link>Utilities\Metadata\AssemblyNameInfo.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\Metadata\TypeName.cs">
-      <Link>Utilities\TypeName.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\Metadata\TypeNameParserOptions.cs">
-      <Link>Utilities\TypeNameParserOptions.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\Metadata\TypeNameParser.cs">
-      <Link>Utilities\TypeNameParser.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\Metadata\TypeNameParserHelpers.cs">
-      <Link>Utilities\TypeNameParserHelpers.cs</Link>
-    </Compile>
     <Compile Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\UnconditionalSuppressMessageAttribute.cs" Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
       <Link>System\Diagnostics\CodeAnalysis\UnconditionalSuppressMessageAttribute.cs</Link>
     </Compile>

--- a/src/coreclr/tools/ILVerify/ILVerify.csproj
+++ b/src/coreclr/tools/ILVerify/ILVerify.csproj
@@ -17,6 +17,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
     <PackageReference Include="System.CommandLine" Version="$(SystemCommandLineVersion)" />
   </ItemGroup>
 </Project>

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/TypeNameParser.Dataflow.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/TypeNameParser.Dataflow.cs
@@ -9,6 +9,8 @@ namespace System.Reflection
 {
     internal partial struct TypeNameParser
     {
+        private static readonly TypeNameParseOptions s_typeNameParseOptions = new() { MaxNodes = int.MaxValue };
+
         private TypeSystemContext _context;
         private ModuleDesc _callingModule;
         private List<ModuleDesc> _referencedModules;
@@ -17,7 +19,7 @@ namespace System.Reflection
         public static TypeDesc ResolveType(string name, ModuleDesc callingModule,
             TypeSystemContext context, List<ModuleDesc> referencedModules, out bool typeWasNotFoundInAssemblyNorBaseLibrary)
         {
-            if (!TypeName.TryParse(name, out TypeName parsed))
+            if (!TypeName.TryParse(name, out TypeName parsed, s_typeNameParseOptions))
             {
                 typeWasNotFoundInAssemblyNorBaseLibrary = false;
                 return null;

--- a/src/coreclr/tools/aot/ILCompiler.TypeSystem/ILCompiler.TypeSystem.csproj
+++ b/src/coreclr/tools/aot/ILCompiler.TypeSystem/ILCompiler.TypeSystem.csproj
@@ -198,38 +198,11 @@
     <Compile Include="..\..\Common\TypeSystem\Common\Utilities\CustomAttributeTypeNameParser.cs">
       <Link>Utilities\CustomAttributeTypeNameParser.cs</Link>
     </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\HexConverter.cs">
-      <Link>Utilities\HexConverter.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\AssemblyNameFormatter.cs">
-      <Link>Utilities\AssemblyNameFormatter.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\AssemblyNameParser.cs">
-      <Link>Utilities\AssemblyNameParser.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\Metadata\AssemblyNameInfo.cs">
-      <Link>Utilities\AssemblyNameInfo.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\Metadata\TypeName.cs">
-      <Link>Utilities\TypeName.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\Metadata\TypeNameParser.cs">
-      <Link>Utilities\TypeNameParser.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\Metadata\TypeNameParserHelpers.cs">
-      <Link>Utilities\TypeNameParserHelpers.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\Metadata\TypeNameParserOptions.cs">
-      <Link>Utilities\TypeNameParserOptions.cs</Link>
-    </Compile>
     <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Reflection\TypeNameParser.Helpers.cs">
       <Link>Utilities\CustomAttributeTypeNameParser.Helpers</Link>
     </Compile>
     <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Text\ValueStringBuilder.cs">
       <Link>Utilities\ValueStringBuilder.cs</Link>
-    </Compile>
-    <Compile Include="$(LibrariesProjectRoot)\Common\src\System\Text\ValueStringBuilder.AppendSpanFormattable.cs">
-      <Link>Utilities\ValueStringBuilder.AppendSpanFormattable.cs</Link>
     </Compile>
     <Compile Include="..\..\Common\TypeSystem\Common\Utilities\GCPointerMap.Algorithm.cs">
       <Link>Utilities\GCPointerMap.Algorithm.cs</Link>
@@ -739,6 +712,10 @@
 
   <ItemGroup>
     <AdditionalFiles Include="BannedSymbols.txt" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
   </ItemGroup>
 
   <ItemGroup>

--- a/src/libraries/Common/src/System/Reflection/Metadata/AssemblyNameInfo.cs
+++ b/src/libraries/Common/src/System/Reflection/Metadata/AssemblyNameInfo.cs
@@ -21,10 +21,10 @@ namespace System.Reflection.Metadata
     /// It's a more lightweight, immutable version of <seealso cref="AssemblyName"/> that does not pre-allocate <seealso cref="System.Globalization.CultureInfo"/> instances.
     /// </remarks>
     [DebuggerDisplay("{FullName}")]
-#if SYSTEM_PRIVATE_CORELIB
-    internal
-#else
+#if SYSTEM_REFLECTION_METADATA
     public
+#else
+    internal
 #endif
     sealed class AssemblyNameInfo
     {
@@ -182,11 +182,7 @@ namespace System.Reflection.Metadata
         public static AssemblyNameInfo Parse(ReadOnlySpan<char> assemblyName)
             => TryParse(assemblyName, out AssemblyNameInfo? result)
                 ? result!
-#if SYSTEM_REFLECTION_METADATA || SYSTEM_PRIVATE_CORELIB
                 : throw new ArgumentException(SR.InvalidAssemblyName, nameof(assemblyName));
-#else // tools that reference this file as a link
-                : throw new ArgumentException("The given assembly name was invalid.", nameof(assemblyName));
-#endif
 
         /// <summary>
         /// Tries to parse a span of characters into an assembly name.

--- a/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs
+++ b/src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs
@@ -16,10 +16,10 @@ using System.Collections.Immutable;
 namespace System.Reflection.Metadata
 {
     [DebuggerDisplay("{AssemblyQualifiedName}")]
-#if SYSTEM_PRIVATE_CORELIB
-    internal
-#else
+#if SYSTEM_REFLECTION_METADATA
     public
+#else
+    internal
 #endif
     sealed class TypeName
     {

--- a/src/libraries/Common/src/System/Reflection/Metadata/TypeNameParser.cs
+++ b/src/libraries/Common/src/System/Reflection/Metadata/TypeNameParser.cs
@@ -17,7 +17,8 @@ namespace System.Reflection.Metadata
     [DebuggerDisplay("{_inputString}")]
     internal ref struct TypeNameParser
     {
-        private static readonly TypeNameParseOptions _defaults = new();
+        private static readonly TypeNameParseOptions s_defaults = new();
+
         private readonly bool _throwOnError;
         private readonly TypeNameParseOptions _parseOptions;
         private ReadOnlySpan<char> _inputString;
@@ -26,7 +27,7 @@ namespace System.Reflection.Metadata
         {
             _inputString = name;
             _throwOnError = throwOnError;
-            _parseOptions = options ?? _defaults;
+            _parseOptions = options ?? s_defaults;
         }
 
         internal static TypeName? Parse(ReadOnlySpan<char> typeName, bool throwOnError, TypeNameParseOptions? options = default)
@@ -50,7 +51,7 @@ namespace System.Reflection.Metadata
             {
                 if (throwOnError)
                 {
-                    if (recursiveDepth >= parser._parseOptions.MaxNodes)
+                    if (parser._parseOptions.IsMaxDepthExceeded(recursiveDepth))
                     {
                         ThrowInvalidOperation_MaxNodesExceeded(parser._parseOptions.MaxNodes);
                     }
@@ -249,7 +250,7 @@ namespace System.Reflection.Metadata
 
         private bool TryDive(ref int depth)
         {
-            if (depth >= _parseOptions.MaxNodes)
+            if (_parseOptions.IsMaxDepthExceeded(depth))
             {
                 return false;
             }

--- a/src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs
+++ b/src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs
@@ -331,53 +331,64 @@ namespace System.Reflection.Metadata
         }
 
         [DoesNotReturn]
-        internal static void ThrowInvalidOperation_MaxNodesExceeded(int limit) => throw
+        internal static void ThrowArgumentException_InvalidTypeName(int errorIndex)
+        {
+            throw new ArgumentException(SR.Argument_InvalidTypeName, $"typeName@{errorIndex}");
+        }
+
+        [DoesNotReturn]
+        internal static void ThrowInvalidOperation_MaxNodesExceeded(int limit)
+        {
 #if SYSTEM_REFLECTION_METADATA
-            new InvalidOperationException(SR.Format(SR.InvalidOperation_MaxNodesExceeded, limit));
-#else // corelib and tools that reference this file as a link
-            new InvalidOperationException();
+            throw new InvalidOperationException(SR.Format(SR.InvalidOperation_MaxNodesExceeded, limit));
+#else
+            Debug.Fail("Expected to be unreachable");
+            throw new InvalidOperationException();
 #endif
+        }
 
         [DoesNotReturn]
-        internal static void ThrowArgumentException_InvalidTypeName(int errorIndex) => throw
-#if SYSTEM_PRIVATE_CORELIB
-            new ArgumentException(SR.Arg_ArgumentException, $"typeName@{errorIndex}");
-#elif SYSTEM_REFLECTION_METADATA
-            new ArgumentException(SR.Argument_InvalidTypeName, $"typeName@{errorIndex}");
-#else // tools that reference this file as a link
-            new ArgumentException();
-#endif
-
-        [DoesNotReturn]
-        internal static void ThrowInvalidOperation_NotGenericType() => throw
-#if SYSTEM_REFLECTION_METADATA || SYSTEM_PRIVATE_CORELIB
-            new InvalidOperationException(SR.InvalidOperation_NotGenericType);
-#else // tools that reference this file as a link
-            new InvalidOperationException();
-#endif
-
-        [DoesNotReturn]
-        internal static void ThrowInvalidOperation_NotNestedType() => throw
+        internal static void ThrowInvalidOperation_NotGenericType()
+        {
 #if SYSTEM_REFLECTION_METADATA
-            new InvalidOperationException(SR.InvalidOperation_NotNestedType);
-#else // corelib and tools that reference this file as a link
-            new InvalidOperationException();
+            throw new InvalidOperationException(SR.InvalidOperation_NotGenericType);
+#else
+            Debug.Fail("Expected to be unreachable");
+            throw new InvalidOperationException();
 #endif
+        }
 
         [DoesNotReturn]
-        internal static void ThrowInvalidOperation_NoElement() => throw
+        internal static void ThrowInvalidOperation_NotNestedType()
+        {
 #if SYSTEM_REFLECTION_METADATA
-            new InvalidOperationException(SR.InvalidOperation_NoElement);
-#else // corelib and tools that reference this file as a link
-            new InvalidOperationException();
+            throw new InvalidOperationException(SR.InvalidOperation_NotNestedType);
+#else
+            Debug.Fail("Expected to be unreachable");
+            throw new InvalidOperationException();
 #endif
+        }
 
         [DoesNotReturn]
-        internal static void ThrowInvalidOperation_HasToBeArrayClass() => throw
-#if SYSTEM_REFLECTION_METADATA || SYSTEM_PRIVATE_CORELIB
-            new InvalidOperationException(SR.Argument_HasToBeArrayClass);
-#else // tools that reference this file as a link
-            new InvalidOperationException();
+        internal static void ThrowInvalidOperation_NoElement()
+        {
+#if SYSTEM_REFLECTION_METADATA
+            throw new InvalidOperationException(SR.InvalidOperation_NoElement);
+#else
+            Debug.Fail("Expected to be unreachable");
+            throw new InvalidOperationException();
 #endif
+        }
+
+        [DoesNotReturn]
+        internal static void ThrowInvalidOperation_HasToBeArrayClass()
+        {
+#if SYSTEM_REFLECTION_METADATA
+            throw new InvalidOperationException(SR.Argument_HasToBeArrayClass);
+#else
+            Debug.Fail("Expected to be unreachable");
+            throw new InvalidOperationException();
+#endif
+        }
     }
 }

--- a/src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserOptions.cs
+++ b/src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserOptions.cs
@@ -3,19 +3,9 @@
 
 namespace System.Reflection.Metadata
 {
-#if SYSTEM_PRIVATE_CORELIB
-    internal
-#else
-    public
-#endif
-    sealed class TypeNameParseOptions
+    public sealed class TypeNameParseOptions
     {
-        private int _maxNodes =
-#if SYSTEM_PRIVATE_CORELIB
-            int.MaxValue; // CoreLib has never introduced any limits
-#else
-            20;
-#endif
+        private int _maxNodes = 20;
 
         /// <summary>
         /// Limits the maximum value of <seealso cref="TypeName.GetNodeCount">node count</seealso> that parser can handle.
@@ -37,5 +27,7 @@ namespace System.Reflection.Metadata
                 _maxNodes = value;
             }
         }
+
+        internal bool IsMaxDepthExceeded(int depth) => depth >= _maxNodes;
     }
 }

--- a/src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs
+++ b/src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs
@@ -8,6 +8,28 @@ using System.Text;
 
 #nullable enable
 
+#if SYSTEM_PRIVATE_CORELIB
+namespace System.Reflection.Metadata
+{
+    internal struct TypeNameParseOptions
+    {
+        public TypeNameParseOptions() { }
+#pragma warning disable CA1822 // Mark members as static
+        // CoreLib does not enforce any limits
+        public bool IsMaxDepthExceeded(int _) => false;
+        public int MaxNodes
+        {
+            get
+            {
+                 Debug.Fail("Expected to be unreachable");
+                 return 0;
+            }
+        }
+#pragma warning restore CA1822
+    }
+}
+#endif
+
 namespace System.Reflection
 {
     internal partial struct TypeNameParser

--- a/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
+++ b/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
@@ -1424,6 +1424,9 @@
     <Compile Include="$(CommonPath)System\LocalAppContextSwitches.Common.cs">
       <Link>Common\System\LocalAppContextSwitches.Common.cs</Link>
     </Compile>
+    <Compile Include="$(CommonPath)System\HexConverter.cs">
+      <Link>Common\System\HexConverter.cs</Link>
+    </Compile>
     <Compile Include="$(CommonPath)System\HResults.cs">
       <Link>Common\System\HResults.cs</Link>
     </Compile>
@@ -1492,9 +1495,6 @@
     </Compile>
     <Compile Include="$(CommonPath)System\Reflection\Metadata\TypeNameParserHelpers.cs">
       <Link>Common\System\Reflection\Metadata\TypeNameParserHelpers.cs</Link>
-    </Compile>
-    <Compile Include="$(CommonPath)System\Reflection\Metadata\TypeNameParserOptions.cs">
-      <Link>Common\System\Reflection\Metadata\TypeNameParserOptions.cs</Link>
     </Compile>
     <Compile Include="$(CommonPath)System\Runtime\Versioning\NonVersionableAttribute.cs">
       <Link>Common\System\Runtime\Versioning\NonVersionableAttribute.cs</Link>
@@ -1598,11 +1598,6 @@
     <Compile Include="$(MSBuildThisFileDirectory)System\Diagnostics\Tracing\TraceLogging\TypeAnalysis.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Diagnostics\Tracing\TraceLogging\XplatEventLogger.cs" Condition="'$(FeatureXplatEventSource)' == 'true'" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\CompilerServices\QCallHandles.cs" Condition="'$(FeatureNativeAot)' != 'true'" />
-  </ItemGroup>
-  <ItemGroup>
-    <Compile Include="$(MSBuildThisFileDirectory)\..\..\Common\src\System\HexConverter.cs">
-      <Link>Common\System\HexConverter.cs</Link>
-    </Compile>
   </ItemGroup>
   <ItemGroup Condition="'$(TargetsWindows)' == 'true'">
     <Compile Include="$(CommonPath)Interop\Windows\Advapi32\Interop.EncryptDecrypt.cs">

